src/core/update_initial_conditions.jl

```diff
@@ -155,123 +155,70 @@ function calculate_ic_quantity(
 end
 
 ############################# Initial Conditions Initialization ############################
+
+function add_initial_condition!(
+    optimization_container::OptimizationContainer,
+    devices::Union{Vector{T}, IS.FlattenIteratorWrapper{T}},
+    ::D,
+    initial_conditions_type::Type{<:InitialConditionType},
+) where {T <: PSY.Component, D <: AbstractDeviceFormulation}
+    _make_initial_conditions!(
+        optimization_container,
+        devices,
+        D(),
+        nothing,
+        ICKey(initial_conditions_type, T),
+        _get_variable_initial_value,
+    )
+end
+
+function add_initial_condition!(
+    optimization_container::OptimizationContainer,
+    devices::Union{Vector{T}, IS.FlattenIteratorWrapper{T}},
+    ::D,
+    initial_conditions_type::Type{<:InitialConditionType},
+    variable_type::Type{<:VariableType},
+) where {T <: PSY.Component, D <: AbstractDeviceFormulation}
+    _make_initial_conditions!(
+        optimization_container,
+        devices,
+        D(),
+        variable_type(),
+        ICKey(initial_conditions_type, T),
+        _get_variable_initial_value,
+    )
+end
+
 function _make_initial_conditions!(
     optimization_container::OptimizationContainer,
     devices::Union{IS.FlattenIteratorWrapper{T}, Vector{T}},
     device_formulation::Union{AbstractDeviceFormulation, AbstractServiceFormulation},
     variable_type::Union{Nothing, VariableType},
     key::ICKey,
-    make_ic_func::Function, # Function to make the initial condition object
     get_val_func::Function, # Function to get the value from the device to intialize
     cache = nothing,
 ) where {T <: PSY.Component}
     length_devices = length(devices)
     parameters = model_has_parameters(optimization_container)
     ic_container = get_initial_conditions(optimization_container)
-    if !has_initial_conditions(ic_container, key)
+    if !haskey(ic_container, key)
         @debug "Setting $(get_entry_type(key)) initial conditions for all devices $(T) based on system data" _group =
             LOG_GROUP_INITIAL_CONDITIONS
         ini_conds = Vector{InitialCondition}(undef, length_devices)
-        set_initial_conditions!(ic_container, key, ini_conds)
+        ic_container[key] = ini_conds
```

Comment:
You might want to use the accessor functions defined in optimization container. One reason is that you have a debug log statement whenever a set occurs. You defeat that by performing a set here.

Reply:
we probably want to clean up correctly the containers for IC's too